### DISCUSSION TITLE
Increased Compatibility with Shallow Metagenomics

### DISCUSSION THREAD

**Discussion by LeeBergstrand** [metagenome-atlas/atlas#455] [Ideas]
I do a lot of work with shallow metagenomes (500,000 - a few million reads) where you don't get a lot of MAGS but lots of contigs. At scale, you can do shallow metagenomics (500,000 reads for ~ human gut-level species complexity) for twice the price of 16S rRNA gene seq per sample while retaining the added benefit of being able to get some functional information. If one gets mags from such datasets, they usually represent the most abundant species in the sample. My current workflow is as follows:

1. Run ATLAS assemble to step `merge_pairs` of `assemble.smk` (or edit the atlas code so that atlas doesn't delete the QCed and Merged Reads)
2. Rejoin the merged and remaining unmerged reads per sample.
3. Taxonomy classify these reads using [Kaiju](https://kaiju.binf.ku.dk) to generate genus abundances.
3. Predict protein fragments from these reads using [FragGeneScanPlusPlus](https://github.com/unipept/FragGeneScanPlusPlus)  (with settings for QCed reads)
5. Run custom HMMs against the protein fragments for functional abundance.

I also have a custom python data-sci stack microbiome analysis library (in RnD at the moment), which I use for making charts from the Kaiju results.

I would be interested in adding some greater compatibility with shallow metagenomics to ATLAS. If you think this type of functionality is out of scope, refactoring the pre-assembly QC steps into their own rule and making it easier to isolate the merged QCed reads for a later independent analysis.

**Comment by LeeBergstrand** [comment]
@SilasK Based on https://github.com/metagenome-atlas/atlas/discussions/423 (Quantification of genes present in the gene catalogue among samples), I see that you have added a feature to generate abundances for genes in the gene catalogue as called from the contigs, correct? In your estimation, how do these abundances compare to those predicted from running HMMs across the raw reads?

**Reply by LeeBergstrand** [reply]
@SilasK Does ATLAS do any analysis where it calculates the abundance of individual contigs? I would be interested in finding a way to taxonomy classify individual contigs using KAIJU and then multiply those annotations by the abundance of individual contigs as a replacement for the RAW READ method.

KAIJU also has the ability to scan and classify individual protein's taxonomy as well so I can run it across the genes or proteins in the gene catalogue.

**Reply by LeeBergstrand** [reply]
We could also use MMSEQS2 for taxonomy classification as well.

**Reply by SilasK** [reply]
Mmseqs taxonomy for genes/contigs is about to be implemented in the eukarote branch.

I don't know how geneprediction from raw reads compare to quantified genes? Do you mean fragescan. 
I always presuposed that assembly + prodigal would be better than fragescan. But maybe I'm ignoring the advancment of these tools. 

Also mapping reads to genes is not straight forward, as you can have one read mapping to two ajecent genes.

**Reply by SilasK** [reply]
Analyses of shallow metagenomics need references, don't you simple use kraken, mOTUs or so?

**Reply by LeeBergstrand** [reply]
> Analyses of shallow metagenomics need references, don't you simply use Kraken, mOTUs or so?

I am currently using [Kaiju](https://kaiju.binf.ku.dk) for this task which basically a tool in the same class of tools as Kraken. However, it works at a protein level rather than a nucleotide level. Kaiju uses the NCBI Non-redundant protein database as a reference, just like how Kracken uses the NR nucleotide database as its reference. Because Kaiju works at a protein level and allows for greedy matches, I find that it works a bit better for analyzing environmental datasets, when compared to Kracken.

> I don't know how geneprediction from raw reads compare to quantified genes? Do you mean fragescan.

For raw read analysis, I think you can go either way. Some tools like Kracken, which uses kmers, can identify taxonomy relevant kmer patterns in reads which are not protein-coding (for example promoter sites and RNA enzyme coding genes). I think using fraggenescan would miss these as they would be skipped by the gene predictor. When I do use raw read taxonomy assignment tools, I usually run them reads that are error corrected and merged by Atlas.

If I'm doing protein functional analysis I use fraggenescan and then run HMMs against the predicted fragments.

**Reply by LeeBergstrand** [reply]
>KAIJU also has the ability to scan and classify individual protein's taxonomy as well so I can run it across the genes or proteins in the gene catalogue.

@SilasK Normally, when you run Kaiju you run it on the nucleotide reads directly. Internally, Kaiju then brute force translates each read to all six reading frames and then uses those protein inputs. It also has a `-p` flag that allows it to accept protein sequences directly.

**Reply by LeeBergstrand** [reply]
This is a pretty good review article, comparing the majority of raw read classifiers including kaiju, kracken, mmseqs2 etc. They made a uniform reference database for all the tools used to remove database bias.

https://doi.org/10.1016/j.cell.2019.07.010

**Reply by LeeBergstrand** [reply]
![image](https://user-images.githubusercontent.com/5819462/138388860-547c1548-bae9-4d74-9cb2-b305f252ddf3.jpeg)

Figure 3. Benchmark AUPR Scores

(A) Area under the precision-recall curve (AUPR) scores for each classifier at the species level (a higher value is better). Each plot point represents the score for a (classifier, dataset combination). Classifiers are grouped and colored by their target class.
(B) AUPR for the uniform RefSeq CG database instead of default databases. Missing entries on the RefSeq CG plot are classifiers that cannot create custom databases.

**Reply by LeeBergstrand** [reply]
![image](https://user-images.githubusercontent.com/5819462/138389183-bf584f45-0154-4229-9de6-b68fb92baf23.jpeg)

Figure 5. Proportion of Abundance Classified at the Species Rank

(A) Proportion of sample abundance classified at the species rank with default databases.
(B) Using uniform RefSeq CG databases. Only programs allowing custom databases are shown.

**Reply by LeeBergstrand** [reply]
![image](https://user-images.githubusercontent.com/5819462/138389379-7d69952f-ce96-45be-82ae-730c492704d9.jpeg)

Figure 7. Benchmark of Computational Resources

(A) Time required to process a sample containing 5.7 million reads versus a second run immediately after the first. This second run is faster for many classifiers because sample reads and database files are cached in memory. Bracken is not plotted because it requires negligible time and memory.

(B) The maximum memory utilized by each classifier during execution, the on-disk database size, and average number of CPUs utilized of 32 available.

(C) Time taken and memory used to create the RefSeq CG database using various methods. Classifiers are sorted by increasing time taken. MMseqs2 and DIAMOND do not index the genomes during database construction but, rather, index on the fly during sample classification.

**Reply by LeeBergstrand** [reply]
> Also mapping reads to genes is not straight forward, as you can have one read mapping to two ajecent genes.

Is this calculation without replacement? i.e., is it only counted toward one gene or the other? Would you want it to count towards both genes? It might also make sense to use the counts of the contigs on which the gene is found in the abundance calculation?

**Comment by LeeBergstrand** [comment]
@SilasK I think the best approach would be to:

1) Use the contig genecatalogue for functional profiling and get functional abundances from either read to gene or read to contig mapping (i.e., how abundant is the contig that gene is on).
2) Taxonomy classify the contigs using mmseqs2 or another tool* and use read-to-contig mapping to get abundances. The output from this could be a [biom](https://biom-format.org) file.

*According to the charts above kaiju might be slightly more accurate and faster than mmseqs2 for taxonomy classification? I think the authors processing speed data might actually be inaccurate for mmseqs2 as they don't take into account that mmseqs2 may be able to prebuild indexes. Can mmseqs2 prebuild taxonomy indexes?

Kaiju has an LCA algorithm as mmseqs2 does so I think you could also potentially use it on contigs instead of reads.

**Comment by LeeBergstrand** [comment]
@SilasK How would you take into account the number of reads which map to the contigs? I assume if you have a lower number of reads mapping to the contigs then you might get lower accuracy of the abundance counts. For example, if one sample had 40% of the reads map to its contigs versus another sample had 80% of the reads map to its contigs, are these two samples comparable?

**Comment by LeeBergstrand** [comment]
Below is a comparison for scanning both the merged reads and the gene catalogue for the same HMMs. Figures have been sanitized to remove sensitive information.

![bokeh_plot-9](https://user-images.githubusercontent.com/5819462/138405538-43c0c8cd-b8ca-412b-8d84-306bfc1a3119.png)

For the figure above ATLAS was stopped at rule `merge_reads` with the snakemake flag the keep temp files on. The resulting QCed, extended, and merged reads, along with reads that could not be merged, were converted to FASTA format and scanned with Fraggenescan++ (with error correction turned off). The resulting protein fragments were scanned for functional proteins using custom HMMs. Hmmsearch cut-off parameters were custom-tuned to work with short reads (effectively loosening them). The figure above show the number of passing HMM hits.

![bokeh_plot-10](https://user-images.githubusercontent.com/5819462/138405553-fdc38a96-7234-42dd-84c5-48311bd521a2.png)

The ATLAS run was continued, and the same HMMs were scanned against the resulting gene catalogue. The same HMM cutoffs were used as would be used for full-length proteins. Proteins in the catalogue that had passing hits for an HMM were multiplied by the number of reads mapped to the gene in the catalogue by the `combine_gene_coverages` command (see https://github.com/metagenome-atlas/atlas/issues/276). The above figure shows the data that was the result. 

As you can see you get pretty similar patterns between the read level results and the gene catalogue results. I think the increased abundances in the gene catalogue results may be a result of being able to use longer input data.

**Comment by LeeBergstrand** [comment]
The paper discussed in https://github.com/metagenome-atlas/atlas/discussions/460 may affect our approach. For example, it recommends not using the gene catalogue genes for taxonomy assignment.

**Reply by SilasK** [reply]
For taxonomy wouldn't you use mOTU or metaphlan?

**Comment by LeeBergstrand** [comment]
@SilasK This tool might be relevant to this discussion: https://github.com/rdpstaff/Xander_assembler

Xander: employing a novel method for efficient gene-targeted metagenomic assembly 

`Xander combines gene assignment with the rapid assembly of full-length or near full-length functional genes from metagenomic data without requiring bulk assembly or post-processing to find genes of interest. HMMs used for assembly can be tailored to the targeted genes, allowing flexibility to improve annotation over generic annotation pipelines.`


https://doi.org/10.1186/s40168-015-0093-6
